File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        size = len(height)
        leftmax_array, rightmax_array = [0] * size, [0] * size
        leftmax_array[0] = height[0]
        for i in range(1,size):
            leftmax_array[i] = max(leftmax_array[i-1],height[i])
        rightmax_array[size - 1] = height[size - 1]
        for i in range(size - 2,0,-1):
            rightmax_array[i] = max(rightmax_array[i + 1],height[i])
        value = 0
        count = 0
        for i in range(size):
            if i == 0 or i == size - 1:
                continue
            value = min(leftmax_array[i - 1],rightmax_array[i + 1])
            if value >= height[i]:
                count += (value - height[i])    

        return count

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int answer = 0;
        int size = height.length;
        int i = 0;
        int j = size - 1;
        int lmax = height[0];
        int rmax = height[size - 1];

        while(i <= j){
            if(lmax <= rmax){
                lmax = Math.max(lmax, height[i]);
                answer += lmax - height[i];
                i++;
            }else{
                rmax = Math.max(rmax, height[j]);
                answer += rmax - height[j];
                j--;
            }
        }
        return answer;
    }
}